python/sanlock.c

```diff
@@ -217,55 +217,24 @@ set_value_error(const char* format, PyObject* obj)
 static PyObject *
 __hosts_to_list(struct sanlk_host *hss, int hss_count)
 {
-    int i, rv;
-    PyObject *ls_list = NULL, *ls_entry = NULL, *ls_value = NULL;
+    int i;
+    PyObject *ls_list = NULL, *ls_entry = NULL;
 
     /* prepare the dictionary holding the information */
     if ((ls_list = PyList_New(0)) == NULL)
         goto exit_fail;
 
     for (i = 0; i < hss_count; i++) {
-        if ((ls_entry = PyDict_New()) == NULL)
-            goto exit_fail;
-
-        /* fill the dictionary information: host_id */
-        if ((ls_value = PyInt_FromLong(hss[i].host_id)) == NULL)
-            goto exit_fail;
-        rv = PyDict_SetItemString(ls_entry, "host_id", ls_value);
-        Py_DECREF(ls_value);
-        if (rv != 0)
-            goto exit_fail;
-
-        /* fill the dictionary information: generation */
-        if ((ls_value = PyInt_FromLong(hss[i].generation)) == NULL)
-            goto exit_fail;
-        rv = PyDict_SetItemString(ls_entry, "generation", ls_value);
-        Py_DECREF(ls_value);
-        if (rv != 0)
-            goto exit_fail;
 
-        /* fill the dictionary information: timestamp */
-        if ((ls_value = PyInt_FromLong(hss[i].timestamp)) == NULL)
-            goto exit_fail;
-        rv = PyDict_SetItemString(ls_entry, "timestamp", ls_value);
-        Py_DECREF(ls_value);
-        if (rv != 0)
-            goto exit_fail;
-
-        /* fill the dictionary information: io_timeout */
-        if ((ls_value = PyInt_FromLong(hss[i].io_timeout)) == NULL)
-            goto exit_fail;
-        rv = PyDict_SetItemString(ls_entry, "io_timeout", ls_value);
-        Py_DECREF(ls_value);
-        if (rv != 0)
-            goto exit_fail;
-
-        /* fill the dictionary information: flags */
-        if ((ls_value = PyInt_FromLong(hss[i].flags)) == NULL)
-            goto exit_fail;
-        rv = PyDict_SetItemString(ls_entry, "flags", ls_value);
-        Py_DECREF(ls_value);
-        if (rv != 0)
+        /* fill the dictionary information */
+        ls_entry = Py_BuildValue(
+                "{sKsKsKsIsI}",
```

Comment:
Doc/extending/extending.rst mention a more readable syntax:

    "{s:K,s:K,s:K,s:I,s:I}"

See https://github.com/python/cpython/blob/2.7/Doc/extending/extending.rst#building-arbitrary-values

Turns out that whitespace is ignored, and this works:

    "{s:K, s:K, s:K, s:I, s:I}"

py2:
https://github.com/python/cpython/blob/951af2d7f140be7beb9cda2bcdd54f820c905e45/Python/modsupport.c#L489

py3.6:
https://github.com/python/cpython/blob/2b9d7abdbd4b41e2c624858f5bc80da59d8a681d/Python/modsupport.c#L414

There is a risk in using poorly documented syntax, so I think it will be good idea to ask about this in python mailing list.

Reply:
seems legit (taken from https://docs.python.org/3/extending/extending.html):

You may also call a function with keyword arguments by using PyObject_Call(), which supports arguments and keyword arguments. As in the above example, we use Py_BuildValue() to construct the dictionary.

PyObject *dict;
...
dict = Py_BuildValue("{s:i}", "name", val);
result = PyObject_Call(my_callback, NULL, dict);
Py_DECREF(dict);
if (result == NULL)
    return NULL; /* Pass error back */
/* Here maybe use the result */
Py_DECREF(result);

Reply:
i think we can agree on "{s:K,s:K,s:K,s:I,s:I}" as its part of the doc for now.